File: 2_Data_Structures/week2_priority_queues_and_disjoint_sets/3_merging_tables/merging_tables.cpp
#include <cstdio>
#include <cstdlib>
#include <vector>
#include <algorithm>
#include <iostream>

using std::cin;
using std::cout;
using std::endl;
using std::max;
using std::vector;

struct DisjointSetsElement {
	int size, parent, rank;

	DisjointSetsElement(int size = 0, int parent = -1, int rank = 0) :
		size(size), parent(parent), rank(rank) {}
};

struct DisjointSets {
	int size;
	int max_table_size;
	vector <DisjointSetsElement> sets;

	DisjointSets(int size) : size(size), max_table_size(0), sets(size) {
		for (int i = 0; i < size; i++)
			sets[i].parent = i;
	}

	//int getParent(int table) {
	//	// find parent and compress path
	//}

	//void merge(int destination, int source) {
	//	int realDestination = getParent(destination);
	//	int realSource = getParent(source);
	//	if (realDestination != realSource) {
	//		// merge two components
	//		// use union by rank heuristic
	//		// update max_table_size
	//	}
	//}

	int getParent(int table) {
		// find parent and compress path
		if (table != sets[table].parent)
			sets[table].parent = getParent(sets[table].parent);
		return sets[table].parent;
	}

	void merge(int destination, int source) {
		int realDestination = getParent(destination);
		int realSource = getParent(source);
		if (realDestination != realSource) {
			// merge two components
			// use union by rank heuristic
			// update max_table_size
			auto& src = sets[realSource];
			auto& dst = sets[realDestination];
			if (src.rank > dst.rank) {
				dst.parent = realSource;
				src.size += dst.size;
				dst.size = 0;
				max_table_size = max(max_table_size, src.size);
			}
			else {
				src.parent = realDestination;
				dst.size += src.size;
				src.size = 0;
				max_table_size = max(max_table_size, dst.size);
				if (dst.rank == src.rank)
					dst.rank++;
			}
		}
	}
};

int main() {
	int n, m;
	cin >> n >> m;

	DisjointSets tables(n);
	for (auto& table : tables.sets) {
		cin >> table.size;
		tables.max_table_size = max(tables.max_table_size, table.size);
	}

	for (int i = 0; i < m; i++) {
		int destination, source;
		cin >> destination >> source;
		--destination;
		--source;

		tables.merge(destination, source);
		cout << tables.max_table_size << endl;
	}

	return 0;
}

File: 2_Data_Structures/week2_priority_queues_and_disjoint_sets/3_merging_tables/MergingTables.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.PrintWriter;
import java.util.Locale;
import java.util.StringTokenizer;

public class MergingTables {
    private final InputReader reader;
    private final OutputWriter writer;

    public MergingTables(InputReader reader, OutputWriter writer) {
        this.reader = reader;
        this.writer = writer;
    }

    public static void main(String[] args) {
        InputReader reader = new InputReader(System.in);
        OutputWriter writer = new OutputWriter(System.out);
        new MergingTables(reader, writer).run();
        writer.writer.flush();
    }

    // Ported from C++ templates
    class DisjointSets {
        class DisjointSetsElement {
            int size, parent, rank;

            public DisjointSetsElement() {
                size = 0;
                parent = -1;
                rank = 0;
            }
        }

        int size;
        int max_table_size;
        DisjointSetsElement[] sets;

        DisjointSets(int size) {
            this.size = size;
            max_table_size = 0;
            sets = new DisjointSetsElement[size];
            for (int i = 0; i < sets.length; ++i) {
                sets[i] = new DisjointSetsElement();
                sets[i].parent = i;
            }
        }

        int getParent(int table) {
            if (table != sets[table].parent)
                sets[table].parent = getParent(sets[table].parent);
            return sets[table].parent;
        }

        void merge(int destination, int source) {
            int pDest = getParent(destination);
            int pSource = getParent(source);
            if (pDest != pSource) {
                DisjointSetsElement src = sets[pSource];
                DisjointSetsElement dst = sets[pDest];
                if (src.rank > dst.rank) {
                    dst.parent = pSource;
                    src.size += dst.size;
                    dst.size = 0;
                    max_table_size = Math.max(max_table_size, src.size);
                } else {
                    src.parent = pDest;
                    dst.size += src.size;
                    src.size = 0;
                    max_table_size = Math.max(max_table_size, dst.size);
                    if (dst.rank == src.rank)
                        dst.rank++;

                }
            }
        }
    }

    public void run() {
        int n = reader.nextInt();
        int m = reader.nextInt();
        DisjointSets tables = new DisjointSets(n);
        for (int i = 0; i < n; i++) {
            int numberOfRows = reader.nextInt();
            tables.sets[i].size = numberOfRows;
            tables.max_table_size = Math.max(tables.max_table_size, numberOfRows);
        }
        for (int i = 0; i < m; i++) {
            int destination = reader.nextInt() - 1;
            int source = reader.nextInt() - 1;
            tables.merge(destination, source);
            writer.printf("%d\n", tables.max_table_size);
        }
    }

    static class InputReader {
        public BufferedReader reader;
        public StringTokenizer tokenizer;

        public InputReader(InputStream stream) {
            reader = new BufferedReader(new InputStreamReader(stream), 32768);
            tokenizer = null;
        }

        public String next() {
            while (tokenizer == null || !tokenizer.hasMoreTokens()) {
                try {
                    tokenizer = new StringTokenizer(reader.readLine());
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }
            return tokenizer.nextToken();
        }

        public int nextInt() {
            return Integer.parseInt(next());
        }

        public double nextDouble() {
            return Double.parseDouble(next());
        }

        public long nextLong() {
            return Long.parseLong(next());
        }
    }

    static class OutputWriter {
        public PrintWriter writer;

        OutputWriter(OutputStream stream) {
            writer = new PrintWriter(stream);
        }

        public void printf(String format, Object... args) {
            writer.print(String.format(Locale.ENGLISH, format, args));
        }
    }
}
